simulation/environment/CacheAPI.py

```diff
@@ -0,0 +1,503 @@
+                    " valid. \n")
+
+            elif response['status'] == "OVER_QUERY_LIMIT":
+                sys.stderr.write("Error: Requester has exceeded quota\n")
+
+            elif response['status'] == "REQUEST_DENIED":
+                sys.stderr.write("Error: API could not complete the request\n")
+
+    return elevations
+
+
+def calculate_time_zones(coords, race):
+    """
+
+    Takes in an array of coordinates, return the time zone relative to UTC, of each location in seconds
+
+    :param np.ndarray coords: (float[N][lat lng]) array of coordinates
+    :param str race: either "FSGP" or "ASC" as a string
+    :returns np.ndarray time_diff: (float[N]) array of time differences in seconds
+
+    """
+
+    timezones_return = np.zeros(len(coords))
+
+    tf = TimezoneFinder()
+
+    if race == "FSGP":
+        # this is when FSGP 2021 starts
+        dt = datetime.datetime(*FSGP.date)
+    else:
+        # this is when ASC 2021 starts
+        dt = datetime.datetime(*ASC.date)
+
+    with tqdm(total=len(coords), file=sys.stdout, desc="Calculating Time Zones") as pbar:
+        for index, coord in enumerate(coords):
+            pbar.update(1)
+            tz_string = tf.timezone_at(lat=coord[0], lng=coord[1])
+            timezone = pytz.timezone(tz_string)
+
+            timezones_return[index] = timezone.utcoffset(dt).total_seconds()
+
+    return timezones_return
+
+
+# ------------------- Weather API -------------------
+def cache_weather(race):
+    """
+
+    Makes calls to Weather API for a given race and caches the results to a .npz file
+
+    :param str race: either "FSGP" or "ASC"
+
+    """
+    print("Different weather data requested and/or weather file does not exist. "
+          "Calling OpenWeather API and creating weather save file...\n")
+
+    # Get coords for race
+    if race == "FSGP":
+        # Get path/coords from cached file
+        origin_coord, dest_coord, coords, waypoints = get_fsgp_coords()
+        coords = np.array([coords[0], coords[-1]])
+
+        weather_file = weather_directory / "weather_data_FSGP.npz"
+    else:
+        # Get path/coords from cached file
+        route_file = route_directory / "route_data.npz"
+
+        if os.path.isfile(route_file):  # check there is a cached gis file
+            with np.load(route_file) as gis_data:
+                coords = gis_data['path']
+                origin_coord = gis_data['origin_coord']
+                dest_coord = gis_data['dest_coord']
+                waypoints = gis_data['waypoints']
+                coords = coords[::constants.REDUCTION_FACTOR]
+
+        else:  # no cached file found -> get coords
+            origin_coord, dest_coord, coords, waypoints = get_asc_coords()
+            coords = coords[::constants.REDUCTION_FACTOR]
+
+        weather_file = weather_directory / "weather_data.npz"
+
+    weather_forecast = update_path_weather_forecast(coords, WEATHER_FREQ, int(WEATHER_DURATION))
```

Comment:
Instead of using a constant `WEATHER_DURATION`, can we get this from the config? Also, maybe store `WEATHER_FREQ` in config, too.